### DISCUSSION TITLE
2D input to create molecular grid image

### DISCUSSION THREAD

**Discussion by bertiewooster** [rdkit/rdkit#5917] [Ideas]
This post is to propose a two-dimensional (nested) list input option to create molecular grid images, and solicit feedback on how to proceed. I'd be happy to tackle any coding can be done in Python.

### Motivation
Using [`Draw.MolsToGridImage`](https://www.rdkit.org/docs/GettingStartedInPython.html#drawing-molecules), I have been creating functions where each row has some meaning, for example a [molecule and its groups off a core](https://bertiewooster.github.io/2022/12/25/RDKit-Find-Groups-Off-Common-Core.html), or [level of fragmentation in mass spectrometry](https://bertiewooster.github.io/2023/01/02/Mass-Spectrometry-Fragmentation-Tree.html):

*Annotated to explain columns*
  
![pyridinols-MCS-groups-annotated](https://user-images.githubusercontent.com/1335522/210288966-789180b4-df1a-4cdd-9705-2c05b11c9f6e.png)

<details>
  <summary>Mass spectrometry fragmentation tree</summary>

*Annotated to explain rows*
  
![mass_spec_frag_tree_CAIr_annotated](https://user-images.githubusercontent.com/1335522/210288889-e53630cc-41eb-41d3-afba-ab4022eeab09.png)

</details>

As a conceptual example, the two-dimensional list of molecules `[[molA, molB], [molC, null_mol, molD]]`, where `null_mol` is a molecule with no atoms (its SMILES string is `''`), produces a molecular grid image like:

```
molA molB
molC      molD
```

### High-level algorithm
The general procedure I have used is:

1. Create a two-dimensional list, where each top-level list element represents a row in the molecular grid image, and each element of a sub-lists represents a molecule. Each row might have a different length, depending on how many molecules are in that row.
2. Note the length of the longest sub-list.
3. Pad each sub-list so it is the length of the longest sub-list.
4. Flatten the two-dimensional list to a one-dimensional list.
5. Repeat steps 1-5 for the legends, highlightAtomLists, and highlightBondLists.
6. Call `Draw.MolsToGridImage` on the one-dimensional lists, setting `molsPerRow` to the length of the longest sub-list.

I think it would be useful to incorporate into RDKit a way to handle steps 2-4 automatically. You would provide a two-dimensional list (a list of sub-lists), where each sub-list represents a row, and the code would handle creating the molecular grid image. That would make it easy for users to conceive and implement a two-dimensional molecular grid image, where the row might represent anything useful to their application.

### Request feedback on how to proceed
If folks agree, here are some ideas of how to proceed:

1. Create a new function, for example `Draw.Mols2dToGridImage`, which requires a two-dimensional list and does not have a `molsPerRow` parameter, but otherwise has the same inputs and output as `Draw.MolsToGridImage`. The new function would perform steps 2-4 above for `mols`, `legends`, `highlightAtomLists`, and `highlightBondLists`, then call `Draw.MolsToGridImage` with `molsPerRow` equal to the length of the longest sub-list, and pass through any other arguments (`subImgSize`, `useSVG`, `returnPNG`, `**kwargs`).
    1. Advantages: The user would know that they're supposed to supply a nested list, and would not get confused about whether a `molsPerRow` parameter can be supplied. This new function only does one thing: transform 2D input into 1D input that `Draw.MolsToGridImage` can accept.
    2. Disadvantages: A new function would need to be added.
2. "Overload" `Draw.MolsToGridImage` so that it detects whether the input is a nested list. If so, it performs steps 2-4.
    1. Advantages: No new function needs to be added.
    2. Disadvantages: The RDKit would override any supplied `molsPerRow` value. A user might accidentally pass in a nested list, then be confused when the RDKit formats their molecular grid image in an unexpected way. Complexity would be added to the code because there would be two types of inputs, 1D or 2D lists.

I believe option 1 is better.

**Comment by greglandrum** [comment]
@bertiewooster I can definitely see the use of this and think it's a good idea.
I also prefer the idea of adding a new function, but I would go a step further with the API change in order to make it a bit easier for the caller: how about taking a nested list of dicts instead of multiple nested lists?

So the input would be something like
```
[
[ {'mol': mol1, 'legend':'whatever', 'highlightAtoms':[0,1], 'highlightAtomColors':[(1,0,0),(0,0,1)],'highlightAtomRadii':[0.4, 0.4],
     'highlightBonds':[3], 'highlightBondColors':[(1,0,1]]}]
[ {'mol': mol2, 'legend':'row 2'}, {'mol':mol3}]
 ...
]
```

As for the name... how about something like `MolListsToGridImage()` or even `MolGridToImage()` (though that is perhaps inviting confusion).

**Reply by bertiewooster** [reply]
Great, glad this functionality would be useful. Good deal on adding a new function.

### Dictionary input
Regarding using a dictionary-based input, I understand the appeal of having all the data in one structure. My concern is that the API would then be quite different for `MolsToGridImage` (1D) vs. this new 2D function, which could confuse users:

- they might reasonably wonder why the 1D and 2D versions have different input parameter formats, when the 1D version could equally support a dictionary-based input.
- if they transition from the 1D to 2D function (I've gone from 1D to 2D mol grids in the course of developing code), they have to change their data structures.

Also, from the RDKit coding side, we'd have to check whether each molecule (dictionary) has each element--just because one molecule has a `legend` key in its dictionary, there's no guarantee that another molecule does. That's not conceptually difficult, but seems a bit tedious and might create edge cases?

So a few options come to mind:

1. Accept only a dictionary for the 2D function only, noting the concerns above.
2. Accept either a dictionary, or multiple lists, for both 1D and 2D versions. This increases the code and API complexity, and involves making changes to the existing 1D function.
3. Accept only multiple lists in the 2D function, making its API similar to the 1D function.

While a dictionary input is more elegant, my initial suggestion to get the 2D functionality in the RDKit sooner, and avoid confusion about the 1D vs. 2D APIs, would be option 3. We could consider later implementing option 2 by upgrading both functions to accept either a dictionary or multiple lists.

### Other parameters
`highlightAtomColors`, `highlightAtomRadii`, and `highlightBondColors` are not in `MolsToGridImage`. Are you suggesting adding those input parameters to this new 2D function?

**Reply by bertiewooster** [reply]
A modified approach of #2 is to make a separate function that accepts just a dictionary, for both 1D and 2D functions. For example, the 1D version could be named `MolsDictToGridImage()`.

If I don't get any more feedback before I have a chance to start coding this, I'll start with option 3, a function perhaps named `Mols2dToGridImage()` that accepts lists as `MolsToGridImage()` does. Even if we end up going with dictionary inputs, the code I write could be used internally to convert 2D lists to 1D lists:

```mermaid
graph TD;
    A[option 1: 2D dict fn, Mols2dDictToGridImage]-->B[option 3: 2D lists fn, Mols2dToGridImage];
    B-->C[existing 1D list fn, MolsToGridImage];
```

It seems like this new 2D function belongs in `rdkit/Chem/Draw/__init__.py` along with [`MolsToGridImage()`](https://github.com/rdkit/rdkit/blob/master/rdkit/Chem/Draw/__init__.py#L607).

**Reply by greglandrum** [reply]
> Regarding using a dictionary-based input, I understand the appeal of having all the data in one structure. My concern is that the API would then be quite different for `MolsToGridImage` (1D) vs. this new 2D function, which could confuse users:

I'm nor particularly concerned about this. The functions do different things, so it's reasonable for them to take different arguments, and I really think that it's *a lot* easier for the user to construct the single list of dictionaries than the multiple lists of lists. There's a small amount of extra care required while writing the new function (using `dict.get` instead of `dict[]`), but that's effort that only has to be done once. But that's your call.

I do not like the name `Mols2dToGridImage()`: it really doesn't make clear what exactly the input is and it's an alternate meaning of "2D" in a block of code which is already using that (i.e. MolDraw2D). I would prefer something like "MolListsToGridImage"

**Reply by bertiewooster** [reply]
About the name, "Lists" seems like it could apply to the input parameter formats, and might even be misleading if we make this new function accept a dictionary while `MolsToGridImage()` accepts lists. Maybe something that represents the output format, because that's what people will use it for. The two concepts that come to mind are "matrix" (this is a rectangular array of images, where each row has a meaning) and "row" (because each row has a meaning). Ideas:
- `MolsMatrixToGridImage()` or `MolsToMatrixGridImage()` or `MolsToGridImageMatrix()`
- `MolsByRowToGridImage()` or `MolsToGridImageByRow()`
- `MolsRowsToGridImage()` or `MolsToGridImageRows()`
- `MolsRowWiseToGridImage()` or `MolsToGridImageRowWise()`

**Reply by greglandrum** [reply]
I'm +1 on either `MolsMatrixToGridImage()` or `MolsMatrixToImage()`

**Comment by bertiewooster** [comment]
Some questions about tests for code as I prepare to write tests:

1. Does it matter if the tests I create are in rdkit/Chem/Draw/UnitTestDraw.py or rdkit/Chem/Draw/UnitTestIPython.py? Both have tests of `MolsToGridImage()`.
2. It seems like `unittest` is mostly used. Should I stick with that, or is `pytest` an option? (I'm more familiar with `pytest` and know how to use `Coverage.py` with it, though `unittest` seems fairly easy to use.)
3. I found that I can run the unit tests in one file using for example that `python rdkit/Chem/Draw/UnitTestDraw.py`. What is the recommended way to run all Python tests? If I run [`python -m unittest`](https://docs.python.org/3/library/unittest.html#unittest-test-discovery) from the top-level directory, I get an import error:

```
(my-rdkit-env) myusername@MacBook-Air rdkit % python -m unittest                                
E
======================================================================
ERROR: rdkit (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: rdkit
Traceback (most recent call last):
  File "/Users/myusername/opt/anaconda3/envs/my-rdkit-env/lib/python3.10/unittest/loader.py", line 470, in _find_test_path
    package = self._get_module_from_name(name)
  File "/Users/myusername/opt/anaconda3/envs/my-rdkit-env/lib/python3.10/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/Users/myusername/Projects/rdkit/rdkit/__init__.py", line 6, in <module>
    from . import rdBase
ImportError: cannot import name 'rdBase' from partially initialized module 'rdkit' (most likely due to a circular import) (/Users/myusername/Projects/rdkit/rdkit/__init__.py)


----------------------------------------------------------------------
Ran 1 test in 0.000s

FAILED (errors=1)
```

I get a similar error if I run `python -m unittest rdkit/Chem/Draw/UnitTestDraw.py`.

I checked the [HowToContribute wiki](https://github.com/rdkit/rdkit/wiki/HowToContribute), including the presentation [An RDKit-centric intro to git and github](https://github.com/rdkit/UGM_2016/blob/master/Presentations/Landrum_Schneider_GitHub_Git_and_RDKit.pdf), and unfortunately they didn't address these questions.

I'll add the answers to a [Contribute to RDKit page](https://github.com/rdkit/rdkit/discussions/5801#discussioncomment-4248823) so they're available in the future.

**Comment by bertiewooster** [comment]
Good news: I've largely completed the coding, have some tests in place with `unittest`, and am adding more tests as I learn how to parametrize `unittest`. Bad news: I've run into problems setting up a development environment--

### Trouble setting up development environment
Is it possible to set up a development environment for the RDKit on a Mac, specifically with an M2 chip? I tried to follow the instructions How to build from source with Conda: [macOS 10.12 (Sierra): Python 3 environment](https://www.rdkit.org/docs/Install.html#macos-10-12-sierra-python-3-environment) but this sequence of problems occurred:

```
(rdkit-clean) myUserName@MacBook-Air rdkit-clean % make
/Users/myUserName/miniconda3/envs/rdkit-clean/include/boost/python/detail/wrap_python.hpp:57:11: fatal error: 'pyconfig.h' file not found
```

I [learned that meant that I needed to install apt](https://stackoverflow.com/questions/57244655/cannot-build-boost-python-library-fatal-error-pyconfig-h-no-such-file-or-dire#57244837), which I tried:

```
(rdkit-clean) myUserName@MacBook-Air rdkit-clean % sudo apt install python3.10-dev 
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java
```

I installed Java but then got this error:

```
(rdkit-clean) myUserName@MacBook-Air rdkit-clean % sudo apt install python3.10-dev
The operation couldn’t be completed. Unable to locate a Java Runtime that supports apt.
Please visit http://www.java.com for information on installing Java.
```

I learned that ["by default Mac does not support APT commands"](https://stackoverflow.com/questions/58598795/why-cant-i-run-apt-install-on-mac). Is there some solution?

### Containerized development environment?
Alternatively, is it possible to lower the bar for code contributions by simplifying this process, for example by setting up a containerized development environment using e.g. Docker?

**Comment by DavidACosgrove** [comment]
Hi Jeremy,
Getting the build environment right is often the most challenging part of
getting started with RDKit development, in my experience.  I attach the
YAML file for the conda env I use for building, and the associated cmake
command.  I then use the following to build RDKit into the python
environment so defined:

export RDBASE=/Users/david/Projects/RDKit/Master

export PYTHONPATH=$RDBASE; export DYLD_LIBRARY_PATH=$RDBASE/lib; export
DYLD_FALLBACK_LIBRARY_PATH="$RDBASE/lib:$PYROOT/lib"

*RDKit* (make -j 6 install ; find $RDBASE/rdkit -name \*.so -exec
install_name_tool -add_rpath $RDBASE/lib {} \; -print )


I created the conda env with:

mamba create -n RDKit.Build
mamba activate RDKit.Build
mamba install -c anaconda numpy matplotlib
mamba install -c anaconda cmake cairo pillow eigen pkg-config
mamba install -c anaconda boost-cpp boost py-boost


I use mamba rather than conda, but if you're using conda you should just be
able to substitute 'conda' for 'mamba' in the above and remove the '-c
anaconda' bit.  Greg told me this week that you can do it directly with
mamba, if you don't include the py-boost package, in which case you
wouldn't need the '-c anaconda' bit with the mamba command either.  The
py-boost in mamba is included in the boost package, IIRC.  I have not yet
got round to trying this.


I normally do my development on an M1 mac mini, but this set of commands
also works on a new M1 MacBook Pro and an old intel-based MBP on both
Monterey and Ventura.  Your mileage may vary, obviously.  People have
helpfully sent me cmake commands in the past which work for them that I
have struggled with.


Hope this helps,

Dave



On Wed, Jan 18, 2023 at 4:26 AM Jeremy Monat ***@***.***>
wrote:

> Good news: I've largely completed the coding, have some tests in place
> with unittest, and am adding more tests as I learn how to parametrize
> unittest. Bad news: I've run into problems setting up a development
> environment--
> Trouble setting up development environment
>
> Is it possible to set up a development environment for the RDKit on a Mac,
> specifically with an M2 chip? I tried to follow the instructions How to
> build from source with Conda: macOS 10.12 (Sierra): Python 3 environment
> <https://www.rdkit.org/docs/Install.html#macos-10-12-sierra-python-3-environment>
> but this sequence of problems occurred:
>
> (rdkit-clean) ***@***.*** rdkit-clean % make
>
> /Users/myUserName/miniconda3/envs/rdkit-clean/include/boost/python/detail/wrap_python.hpp:57:11: fatal error: 'pyconfig.h' file not found
>
>
> I learned that meant that I needed to install apt
> <https://stackoverflow.com/questions/57244655/cannot-build-boost-python-library-fatal-error-pyconfig-h-no-such-file-or-dire#57244837>,
> which I tried:
>
> (rdkit-clean) ***@***.*** rdkit-clean % sudo apt install python3.10-dev
>
> The operation couldn’t be completed. Unable to locate a Java Runtime.
>
> Please visit http://www.java.com for information on installing Java
>
>
> I installed Java but then got this error:
>
> (rdkit-clean) ***@***.*** rdkit-clean % sudo apt install python3.10-dev
>
> The operation couldn’t be completed. Unable to locate a Java Runtime that supports apt.
>
> Please visit http://www.java.com for information on installing Java.
>
>
> I learned that "by default Mac does not support APT commands"
> <https://stackoverflow.com/questions/58598795/why-cant-i-run-apt-install-on-mac>.
> Is there some solution?
> Containerized development environment?
>
> Alternatively, is it possible to lower the bar for code contributions by
> simplifying this process, for example by setting up a containerized
> development environment using e.g. Docker?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/5917#discussioncomment-4713403>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FUG3Q4R35AWBCK7KWDWS5WF3ANCNFSM6AAAAAATPKDVOQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Comment by DavidACosgrove** [comment]
PS
I re-read my email twice before posting, and still failed to notice that I
included my prompt in one of the lines.  It should say
(make -j 6 install ; find $RDBASE/rdkit -name \*.so -exec install_name_tool
-add_rpath $RDBASE/lib {} \; -print )

on the 3rd line of the build commands.  Sorry.



On Wed, Jan 18, 2023 at 6:58 AM David Cosgrove ***@***.***>
wrote:

> Hi Jeremy,
> Getting the build environment right is often the most challenging part of
> getting started with RDKit development, in my experience.  I attach the
> YAML file for the conda env I use for building, and the associated cmake
> command.  I then use the following to build RDKit into the python
> environment so defined:
>
> export RDBASE=/Users/david/Projects/RDKit/Master
>
> export PYTHONPATH=$RDBASE; export DYLD_LIBRARY_PATH=$RDBASE/lib; export
> DYLD_FALLBACK_LIBRARY_PATH="$RDBASE/lib:$PYROOT/lib"
>
> *RDKit* (make -j 6 install ; find $RDBASE/rdkit -name \*.so -exec
> install_name_tool -add_rpath $RDBASE/lib {} \; -print )
>
>
> I created the conda env with:
>
> mamba create -n RDKit.Build
> mamba activate RDKit.Build
> mamba install -c anaconda numpy matplotlib
> mamba install -c anaconda cmake cairo pillow eigen pkg-config
> mamba install -c anaconda boost-cpp boost py-boost
>
>
> I use mamba rather than conda, but if you're using conda you should just
> be able to substitute 'conda' for 'mamba' in the above and remove the '-c
> anaconda' bit.  Greg told me this week that you can do it directly with
> mamba, if you don't include the py-boost package, in which case you
> wouldn't need the '-c anaconda' bit with the mamba command either.  The
> py-boost in mamba is included in the boost package, IIRC.  I have not yet
> got round to trying this.
>
>
> I normally do my development on an M1 mac mini, but this set of commands
> also works on a new M1 MacBook Pro and an old intel-based MBP on both
> Monterey and Ventura.  Your mileage may vary, obviously.  People have
> helpfully sent me cmake commands in the past which work for them that I
> have struggled with.
>
>
> Hope this helps,
>
> Dave
>
>
>
> On Wed, Jan 18, 2023 at 4:26 AM Jeremy Monat ***@***.***>
> wrote:
>
>> Good news: I've largely completed the coding, have some tests in place
>> with unittest, and am adding more tests as I learn how to parametrize
>> unittest. Bad news: I've run into problems setting up a development
>> environment--
>> Trouble setting up development environment
>>
>> Is it possible to set up a development environment for the RDKit on a
>> Mac, specifically with an M2 chip? I tried to follow the instructions How
>> to build from source with Conda: macOS 10.12 (Sierra): Python 3
>> environment
>> <https://www.rdkit.org/docs/Install.html#macos-10-12-sierra-python-3-environment>
>> but this sequence of problems occurred:
>>
>> (rdkit-clean) ***@***.*** rdkit-clean % make
>>
>> /Users/myUserName/miniconda3/envs/rdkit-clean/include/boost/python/detail/wrap_python.hpp:57:11: fatal error: 'pyconfig.h' file not found
>>
>>
>> I learned that meant that I needed to install apt
>> <https://stackoverflow.com/questions/57244655/cannot-build-boost-python-library-fatal-error-pyconfig-h-no-such-file-or-dire#57244837>,
>> which I tried:
>>
>> (rdkit-clean) ***@***.*** rdkit-clean % sudo apt install python3.10-dev
>>
>> The operation couldn’t be completed. Unable to locate a Java Runtime.
>>
>> Please visit http://www.java.com for information on installing Java
>>
>>
>> I installed Java but then got this error:
>>
>> (rdkit-clean) ***@***.*** rdkit-clean % sudo apt install python3.10-dev
>>
>> The operation couldn’t be completed. Unable to locate a Java Runtime that supports apt.
>>
>> Please visit http://www.java.com for information on installing Java.
>>
>>
>> I learned that "by default Mac does not support APT commands"
>> <https://stackoverflow.com/questions/58598795/why-cant-i-run-apt-install-on-mac>.
>> Is there some solution?
>> Containerized development environment?
>>
>> Alternatively, is it possible to lower the bar for code contributions by
>> simplifying this process, for example by setting up a containerized
>> development environment using e.g. Docker?
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rdkit/rdkit/discussions/5917#discussioncomment-4713403>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACGF2FUG3Q4R35AWBCK7KWDWS5WF3ANCNFSM6AAAAAATPKDVOQ>
>> .
>> You are receiving this because you are subscribed to this thread.Message
>> ID: ***@***.***>
>>
>
>
> --
> David Cosgrove
> Freelance computational chemistry and chemoinformatics developer
> http://cozchemix.co.uk
>
>

-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by bertiewooster** [reply]
Thanks @DavidACosgrove. Here are the commands I ran. (I got the error `find: /Users/jemonat/Projects/RDKit/Master/rdkit: No such file or directory` when I used `export export RDBASE=/Users/david/Projects/RDKit/Master`, perhaps because I don't have a folder for that branch (?), so I removed the `/Master` from the end.)

```
conda create -n RDKit.Build
conda activate RDKit.Build
conda install numpy matplotlib
conda install cmake cairo pillow eigen pkg-config
conda install boost-cpp boost py-boost
export RDBASE=/Users/jemonat/Projects/RDKit
export PYTHONPATH=$RDBASE; export DYLD_LIBRARY_PATH=$RDBASE/lib; export DYLD_FALLBACK_LIBRARY_PATH="$RDBASE/lib:$PYROOT/lib"
(make -j 6 install ; find $RDBASE/rdkit -name \*.so -exec install_name_tool -add_rpath $RDBASE/lib {} \; -print )
```

`make` ran for a while, then gave this error:

```
[ 19%] Building C object External/AvalonTools/CMakeFiles/avalon_clib_static.dir/ava-formake-AvalonToolkit_2.0.4a/src/main/C/common/hashcode.c.o
[ 19%] Linking C shared library ../../lib/libRDKitavalon_clib.dylib
2 warnings generated.
6 warnings generated.
[ 19%] Linking C static library libRDKitavalon_clib_static.a
[ 19%] Built target avalon_clib
[ 19%] Built target avalon_clib_static
[ 19%] Linking CXX executable testDict
[ 19%] Built target testDict
make: *** [all] Error 2
```

A web search didn't return any results that seemed applicable, so I request the community's help.

**Reply by bertiewooster** [reply]
Follow up: I tried the Mamba option too. After I set my PYTHONPATH using `echo PYTHONPATH=/Users/jemonat/mambaforge`, I got somewhat further with `(make -j 6 install ; find $RDBASE/rdkit -name \*.so -exec install_name_tool -add_rpath $RDBASE/lib {} \; -print )`, but a similar error to my first attempt with Conda, which I previously tried to solve by installing Java to no avail:

```
/Users/jemonat/opt/anaconda3/envs/my-rdkit-env/include/boost/python/detail/wrap_python.hpp:57:11: fatal error: 'pyconfig.h' file not found
# include <pyconfig.h>
          ^~~~~~~~~~~~
[ 20%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs_static.dir/SparseBitVect.cpp.o
[ 20%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs_static.dir/ExplicitBitVect.cpp.o
[ 20%] Linking CXX shared library ../../../lib/libRDKitEigenSolvers.dylib
1 error generated.
make[2]: *** [Code/RDBoost/CMakeFiles/RDBoost.dir/Wrap.cpp.o] Error 1
make[1]: *** [Code/RDBoost/CMakeFiles/RDBoost.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
1 error generated.
make[2]: *** [Code/RDBoost/CMakeFiles/RDBoost_static.dir/Wrap.cpp.o] Error 1
make[1]: *** [Code/RDBoost/CMakeFiles/RDBoost_static.dir/all] Error 2
[ 20%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs.dir/ExplicitBitVect.cpp.o
[ 20%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs_static.dir/Utils.cpp.o
[ 20%] Built target EigenSolvers
[ 20%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs.dir/Utils.cpp.o
[ 20%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs_static.dir/base64.cpp.o
[ 20%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs_static.dir/BitOps.cpp.o
[ 20%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs_static.dir/DiscreteDistMat.cpp.o
[ 20%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs.dir/base64.cpp.o
[ 21%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs.dir/BitOps.cpp.o
[ 21%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs.dir/DiscreteDistMat.cpp.o
[ 21%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs_static.dir/DiscreteValueVect.cpp.o
[ 21%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs_static.dir/FPBReader.cpp.o
[ 21%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs.dir/DiscreteValueVect.cpp.o
[ 21%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs.dir/FPBReader.cpp.o
[ 22%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs_static.dir/MultiFPBReader.cpp.o
[ 22%] Building CXX object Code/DataStructs/CMakeFiles/DataStructs.dir/MultiFPBReader.cpp.o
[ 22%] Linking CXX static library libRDKitDataStructs_static.a
[ 22%] Built target DataStructs_static
[ 22%] Linking CXX shared library ../../lib/libRDKitDataStructs.dylib
[ 22%] Built target DataStructs
make: *** [all] Error 2
```

**Reply by bertiewooster** [reply]
Thanks to @DavidACosgrove for helping me get my environment set up. The steps were:

- Start with a fresh clone of my fork of the RDKit repo
- Use conda as the package manager and create a new conda environment
- Create and go into a build directory:
```
mkdir build
cd build
```
- Run this `cmake` command
```
cmake -DPy_ENABLE_SHARED=1 \
      -DBOOST_ROOT="$CONDA_PREFIX" \
      -DBoost_NO_SYSTEM_PATHS=ON \
      -DBoost_NO_BOOST_CMAKE=TRUE \
      -DRDK_INSTALL_INTREE=ON \
      -DRDK_INSTALL_STATIC_LIBS=OFF \
      -DRDK_BUILD_CPP_TESTS=ON \
      -DPYTHON_NUMPY_INCLUDE_PATH="$(python -c 'import numpy ; print(numpy.get_include())')" \
      -DRDK_BUILD_CAIRO_SUPPORT=ON \
      -DRDK_BUILD_FREETYPE_SUPPORT=ON \
      ..
```
- Run this
```
(make -j 6 install ; find $RDBASE/rdkit -name \*.so -exec install_name_tool -add_rpath $RDBASE/lib {} \; -print )
```

**Comment by bertiewooster** [comment]
Noting current code flow as an aid during development:

```mermaid
graph TD;
    A[MolsMatrixToGridImage]-->B[_MolsNestedToLinear];
    B-->C[MolsToGridImage pre-existing];
```
